lib/rack/request.rb

```diff
@@ -516,7 +516,10 @@ def POST
             set_header RACK_REQUEST_FORM_INPUT, nil
             set_header(RACK_REQUEST_FORM_HASH, {})
           elsif form_data? || parseable_data?
-            unless set_header(RACK_REQUEST_FORM_HASH, parse_multipart)
+            if pairs = Rack::Multipart.parse_multipart(env, Rack::Multipart::ParamList)
```

Comment:
I originally thought that we should check if `RACK_REQUEST_FORM_PAIRS` is already set here, and if so, use it instead of calling `Rack::Multipart.parse_multipart`.   However, we don't do that for `RACK_REQUEST_FORM_VARS`.  However, since we can no longer reparse the input, do you think we should make the change for both?

Reply:
I initially wrote it that way, then realised it was wrong: `_HASH` and `_PAIRS` get set at the same time, so if we get here when `_PAIRS` is set, that means we decided we need to re-parse, so we must have a new input: https://github.com/rack/rack/blob/8f5c885f7e0427b489174a55e6d88463173f22d2/lib/rack/request.rb#L506-L510

... which does mean I've arguably propagated an existing misbehaviour here: everywhere that we set `_HASH` should really be clearing `_VARS` [and now `_PAIRS`], for instances where the new input has a different format from the old.

Reply:
Consider:

```ruby
request.POST
request.delete_header(RACK_REQUEST_FORM_HASH)
request.POST
```

No change in input in this case.  I guess you could argue that deleting that header means all bets are off, but it seems that using `RACK_REQUEST_FORM_VARS` or `RACK_REQUEST_FORM_PAIRS` would be better in this case.

This idea probably makes more sense if you consider that @tenderlove, @ioquatix, and I all want to remove the automatic cache invalidation currently implemented (#2073 being one approach for that).